### PR TITLE
[opt](function) Optimize the concat(col, constant, constant, constant) function

### PR DESCRIPTION
**Pull request by Mryange**
## Proposed changes

```
mysql [test]>select count(concat(short, "123","121231","123123",'12312313')) from strings;
+-------------------------------------------------------------+
| count(concat(short, '123', '121231', '123123', '12312313')) |
+-------------------------------------------------------------+
|                                                    10000000 |
+-------------------------------------------------------------+
1 row in set (0.52 sec)

mysql [test]>select count(concat(short, "123","121231","123123",'12312313' , short , short, short)) from strings;
+----------------------------------------------------------------------------------+
| count(concat(short, '123', '121231', '123123', '12312313', short, short, short)) |
+----------------------------------------------------------------------------------+
|                                                                         10000000 |
+----------------------------------------------------------------------------------+
1 row in set (0.98 sec)



now

mysql [test]>select count(concat(short, "123","121231","123123",'12312313')) from strings;
+-------------------------------------------------------------+
| count(concat(short, '123', '121231', '123123', '12312313')) |
+-------------------------------------------------------------+
|                                                    10000000 |
+-------------------------------------------------------------+
1 row in set (0.19 sec)

mysql [test]>select count(concat(short, "123","121231","123123",'12312313' , short , short, short)) from strings;
+----------------------------------------------------------------------------------+
| count(concat(short, '123', '121231', '123123', '12312313', short, short, short)) |
+----------------------------------------------------------------------------------+
|                                                                         10000000 |
+----------------------------------------------------------------------------------+
1 row in set (0.71 sec)
```

### CODE REVIEW

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run performance

File: be/src/vec/functions/function_string.h [hunk 1]
```diff
@@ -1017,6 +1022,34 @@ class FunctionStringConcat : public IFunction {
...
+
+        // Optimize function calls like this:
+        // concat(col, "123", "abc", "456") -> tail = "123abc456"
+        for (size_t i = 1; i < context->get_num_args(); i++) {
+            const auto* column_string = context->get_constant_col(i);
+            if (column_string == nullptr) {
+                state->use_state = false;
+                break;
+            }
+            auto string_vale = column_string->column_ptr->get_data_at(0);
```
Comment (HappenLee) [inline be/src/vec/functions/function_string.h:1043]: logic error， if here is NULL， result is ERROR

Comment (HappenLee) [inline be/src/vec/functions/function_string.h:1043]: Check here is NULL const， what happen

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run external

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by zclllyybb — APPROVED** [review]

**Pull request merged**